src/ucp/stream/stream_recv.c

```diff
@@ -8,132 +8,399 @@
+        return UCS_STATUS_PTR(UCS_ERR_NO_PROGRESS);
+    } else if (ucs_queue_is_empty(&ep_stream->reqs)) {
+        ucs_assert(ucs_queue_is_empty(&ep_stream->reqs));
+        ucs_list_add_tail(&worker->stream_eps, &ep_stream->list);
         ep->flags |= UCP_EP_FLAG_STREAM_IS_QUEUED;
-        ucs_list_add_tail(&worker->stream_eps, &ep->ext.stream->list);
+        ucs_queue_push(&ep_stream->data, &rdesc->stream_queue);
+        return UCS_STATUS_PTR(UCS_ERR_NO_PROGRESS);
     }
 
-    ucs_queue_push(&ep->ext.stream->data, &rdesc->stream_queue);
+    return ep_stream;
+}
 
-out:
-    UCP_THREAD_CS_EXIT_CONDITIONAL(&worker->mt_lock);
-    return status;
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_stream_am_handler(void *am_arg, void *am_data, size_t am_length,
+                      unsigned am_flags)
+{
+    ucp_worker_h        worker      = am_arg;
+    ucp_stream_am_hdr_t *hdr        = am_data;
+    size_t              recv_len    = am_length - sizeof(ucp_stream_am_hdr_t);
+    ucp_ep_ext_stream_t *ep_stream  = NULL;
+    ucp_recv_desc_t     *rdesc;
+    ucp_request_t       *req;
+    void                *rdata_ptr;
+    size_t              rdata_len;
+    ssize_t             unpacked;
+    ucs_status_t        status;
+    int                 is_handled;
+
+    ucs_assert(am_length >= sizeof(ucp_stream_am_hdr_t));
+
+    rdesc = ucp_stream_am_rdesc_get(worker, am_data, am_length, am_flags);
+    if (ucs_unlikely(rdesc == NULL)) {
+        goto out;
+    }
 
-err:
-    if (rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC) {
-        ucs_mpool_put(rdesc);
+    ep_stream = ucp_stream_am_get_ep(worker, hdr->sender_uuid, rdesc);
+    if (!UCS_PTR_IS_PTR(ep_stream)) {
+        goto out;
     }
 
-    status = UCS_OK;
-    goto out;
-}
+    is_handled = 0;
+
+    do {
+        req = ucs_queue_pull_elem_non_empty(&ep_stream->reqs, ucp_request_t,
+                                            recv.queue);
+        do {
+            status = ucp_stream_ep_rdesc_set(ep_stream, rdesc, recv_len,
+                                             is_handled);
+            if (status == UCS_OK) {
+                is_handled = 1;
+            } else if (status == UCS_ERR_NO_MESSAGE) {
+                ucp_request_complete_stream_recv(req, UCS_OK);
+                goto out;
+            }
+
+            rdata_ptr = ucp_stream_rdesc_data_offset(ep_stream->rdesc,
+                                                     ep_stream->rdesc_offset);
+            rdata_len = ep_stream->rdesc_len - ep_stream->rdesc_offset;
+            unpacked  = ucp_stream_data_unpack(rdata_ptr, rdata_len, req);
+
+            if (ucs_unlikely(unpacked < 0)) {
+                goto out;
+            } else if (unpacked < rdata_len) {
+                ep_stream->rdesc_offset += unpacked;
+                ucs_assert(ep_stream->rdesc_offset <= ep_stream->rdesc_len);
+                /* This request is full, try next one */
+                ucp_request_complete_stream_recv(req, UCS_OK);
```

Comment:
Why do you complete the recv req here? What is the condition which says that no more data is intended for this request?

Reply:
Couple lines above:
`} else if (unpacked < rdata_len) {`
this means `rdata_len` less than available space in the request

Reply:
i wonder if it is possible that ep has data and is not picked in worker_poll() because some data were picked by a pending request

Reply:
Also it would be nice to unify request advancing with the ucp_stream_recv_nb()

Reply:
Not completely understand... Common part of request advancing is already done in `ucp_stream_data_unpack`. Maybe `if-else-statements` look similar, but code in these blocks are different.